### PR TITLE
feat: booking process is now fully automated

### PR DESCRIPTION
**Pull request by timoknapp**
Hi @TobseF ,

this new PR extends the current state of this bot to be now also used for final vaccination date bookings.

It works as follows:

Whenever a free slot is available:
   
### 4. Takes screenshots of all steps (will be stored statically: `$HOME/Desktop/`)

These are the required additions within `config.properties`:
```
...
# Automatically selects the first possible vaccination date and enters all required personal data based on the following values to book this vaccination slot for you.
bookingEnabled = false
# Takes screenshots during booking process to have proof of you booking. (Optional)
takeScreenshots = false
# Defines the path were all screenshots will be stored in (Optional). Files will be prefixed with "impf-bot-"
outputPath = /home
# Personal Data
# Salutation (m=man, w=women, d=divers, c=child)
personalDataSalutation = m
# Firstname
personalDataFirstname = Max
# Lastname
personalDataLastname = Mustermann
# Zipcode
personalDataZipcode = 69124
# City
personalDataCity = Heidelberg
# Street
personalDataStreet = Hauptstraße
# House number
personalDataHouseNumber = 911
# Phone number. Numbers after the "+49"
personalDataMobileNumber = 152123123123
# E-Mail
personalDataEmail = impfMePlease@lasthope.de
```

Please let me know if these changes match your expectations. ~~I am still prettifying some minor things and removing some `//TODO`s though..~~

Cheers
Timo

### CODE REVIEW

File: src/main/kotlin/de/tfr/impf/ReportJob.kt [hunk 1]
```diff
@@ -107,13 +121,89 @@ class ReportJob {
...
+        bookingPage.enterPhone(personalDataMobileNumber)
+        log.info { "Phone entered..." }
+        Thread.sleep(2_000)
+        bookingPage.enterEmail(personalDataEmail)
+        log.info { "Email entered..." }
+        Thread.sleep(2_000)
+        bookingPage.submitInputWithText("Übernehmen")
+    }
+
+    private fun takeScreenshot(driver: WebDriver, fileName: String) {
```
Comment (TobseF) [inline src/main/kotlin/de/tfr/impf/ReportJob.kt:204]: Should be optional, or use config, to make it compatible with the docker setup.
What happens when `Desktop` directory is missing?

Comment (timoknapp) [reply src/main/kotlin/de/tfr/impf/ReportJob.kt:204]: I changed that part to be optional and also made the output path changeable through a property `outputPath`. So it should be fine for Docker too.

**Comment by piechade** [comment]
After the booking you can see what vaccination you get. Can you build a filter for it? For example if the vaccination is not Biontech than cancel the appointment automatically.

**Comment by timoknapp** [comment]
> After the booking you can see what vaccination you get. Can you build a filter for it? For example if the vaccination is not Biontech than cancel the appointment automatically.

Hi @piechade,

my changes still require a valid placement code ("Vermittlungscode"). The vaccine type is directly related to the placement code itself. The vaccine type again is strongly coupled to the age which you configure within the bot.

Here is the official list of the vaccines with regards to their age requirements.
https://001-iz.impfterminservice.de/assets/static/its/vaccination-list.json

As you can see in the attached list above you will most likely receive the following types based on your age:
16+ = Biontech
18+ = Biontech / Moderna
60+ = AstraZeneca / Johnson&Johnson (Biontech / Moderna, pretty unlikely)

**Comment by piechade** [comment]
Is this really clear like this? First you have the problem that 18+ has Biontech and Moderna. Second you can also get Biontech as 60+.

**Comment by timoknapp** [comment]
Your second point is technically right, but based on current development it has been pretty unlikely to get a placement code with "L920" (Biontech) with a configured age above 60. At least thats what I experienced throughout the past weeks...

**Comment by robinschneider** [comment]
Would be great to have an option to add a not before date XY.

**Comment by Maxopoly** [comment]
I had already been running this PR for a few days and just wanted to report that it did not work for me. Debug log of a missed available slot is:

```
02.06 10:46:17 DEBUG ReportJob - Changed to location page: 76287 Rheinstetten
02.06 10:47:09 INFO ReportJob - Found free vaccination dates in location 76287 Rheinstetten:https://002-iz.impfterminservice.de/impftermine/suche/ABCD-CENS-ORED-ABCD
Ten minutes left to choose a date.
Selected first vaccindation date
Second booing step is shown
(...  lots of emitted slack debug logging  ...)
02.06 10:47:27 DEBUG ReportJob - Open personal data modal...
02.06 10:48:13 DEBUG ReportJob - Choose State: Bitte auswählen
02.06 10:48:47 DEBUG ReportJob - Waiting in WaitingRoomPage ...
02.06 10:48:50 DEBUG ReportJob - Waiting in WaitingRoomPage ... (x100+)
(...)
02.06 11:00:53 DEBUG ReportJob - Changed to location page: 76530 Baden-Baden
02.06 11:01:46 DEBUG ReportJob - Correct code, we can't see the bookable vaccination dates: 76530 Baden-Baden
02.06 11:02:17 DEBUG ReportJob - Choose State: Bitte auswählen
02.06 11:02:52 DEBUG ReportJob - Changed to location page: 75175 Pforzheim
02.06 11:03:56 DEBUG ReportJob - Correct code, we can't see the bookable vaccination dates: 75175 Pforzheim
02.06 11:04:26 DEBUG ReportJob - Choose State: Bitte auswählen


```

It did properly screenshot the available dates, but from there on seemed to just go back to searching. A friend was able to manually get a few appointments at this location around the time the bot reported available ones, so it should have been a failure in the bot.

**Comment by timoknapp** [comment]
@Maxopoly Do you have any more infos about where in the booking process it exactly fails ?

**Comment by Maxopoly** [comment]
I observed another instance of it happening just now, it selected the first available dates and then got stuck in the screen asking for personal information (Name, Address etc.). It entered no information, threw no error in console and just waited in this window, doing nothing. I presume it'd have just gone back to searching a bit after that (I killed the bot after waiting for it to do anything, so I could manually fill in the data).

**Comment by robinschneider** [comment]
I had the bot running in two instances over night and both got an appointment but the automatic booking failed as well.
Screenshots was enabled but it did none, also I wasn't awake so no info on what was shown on screen.
Debug log was not enabled but it reached the second booking step according to the log I got.

**Comment by timoknapp** [comment]
> I had the bot running in two instances over night and both got an appointment but the automatic booking failed as well.
> Screenshots was enabled but it did none, also I wasn't awake so no info on what was shown on screen.
> Debug log was not enabled but it reached the second booking step according to the log I got.

How did you start the instances? If it was running in an headless environment like a raspberry pi, I expect screenshots not to be working, as there is no real screen from which the screenshot could be taken of.

**Comment by timoknapp** [comment]
> I observed another instance of it happening just now, it selected the first available dates and then got stuck in the screen asking for personal information (Name, Address etc.). It entered no information, threw no error in console and just waited in this window, doing nothing. I presume it'd have just gone back to searching a bit after that (I killed the bot after waiting for it to do anything, so I could manually fill in the data).

Okay thanks for the feedback. I will test it myself and check it out in detail.

**Comment by robinschneider** [comment]
> How did you start the instances? If it was running in an headless environment like a raspberry pi, I expect screenshots not to be working, as there is no real screen from which the screenshot could be taken of.


I had it running on a Windows PC with a monitor.

**Comment by timoknapp** [comment]
> > How did you start the instances? If it was running in an headless environment like a raspberry pi, I expect screenshots not to be working, as there is no real screen from which the screenshot could be taken of.
> 
> I had it running on a Windows PC with a monitor.

Hm okay then. Did you specify the correct output folder? Anyways you must see the related logs whenever the relevant screens appear.

**Comment by timoknapp** [comment]
I just tested it myself and the whole process of entering the data including taking screenshots worked as expected. It is important to properly set the property `outputPath` if required. Otherwise leave it empty it will automatically store the screenshots within your home directory:
 - Mac: `/Users/${your-username}`
 - Windows `C:\Users\${your-username}`

**Comment by prok42** [comment]
If the bot automatically and successfully books an appointment, will it stop trying to book other appointments? Otherwise, could it happen that it cancels the appointment by accident (e.g. if you have only one code for three vaccination sites and one was successful)?

**Comment by timoknapp** [comment]
Hi @prok42 ,

No it will not stop. It will continue trying to book new dates for all other locations specified. I am not sure whether your case would then occur, because the cancellation of bookings will be handled by different UIs. Nevertheless we could extend the behaviour in way, such that it will stop looking for new dates whenever it is successful.
I might add this in a new PR @TobseF 😉

**Pull request merged**